viewing_party/party.py

```diff
@@ -1,23 +1,201 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    if title and genre and rating:
+        print("hello")
+        dict = {
+            "title": title,
+            "genre": genre,
+            "rating": rating
+        }
+        return dict
+
+    return None
+
+def add_to_watched(user_data, movie):
+    user_data_copy = user_data.copy()
+
+    # if movie:
+    user_data_copy["watched"].append(movie)
+
+    return user_data_copy
+
+def add_to_watchlist(user_data, movie):
+    user_data_copy = user_data.copy()
+
+    user_data_copy["watchlist"].append(movie)
+
+    return user_data_copy
+
+def watch_movie(user_data, title):
+    # user_data_copy = user_data.copy()
+    for movie_entry in user_data["watchlist"]:
+        if movie_entry["title"] == title:
+            user_data["watched"].append(movie_entry)
+            user_data["watchlist"].remove(movie_entry)
+
+    return user_data
+
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    sum = 0
+    # edge case
+    if len(user_data["watched"]) == 0:
+        return 0.0
+
+    for movie in user_data["watched"]:
+        sum += movie["rating"]
+
+    return sum / len(user_data["watched"])
+
+def get_most_watched_genre(user_data):
+    # edge case
+    if len(user_data["watched"]) == 0:
+        return None
+
+    genre_frequency = {}
+
+    # iterate over the list of watched movies
+        # if genre key not in dict
+            # new_dict[key] = 1
+        # else
+            # +1
+    for movie in user_data["watched"]:
+        if movie["genre"] not in genre_frequency:
+            genre_frequency[movie["genre"]] = 1
+        else:
+            genre_frequency[movie["genre"]] += 1
+
+    # print(genre_frequency)
+    # return key that has max value in genre frequency
+    return max(genre_frequency, key=genre_frequency.get)
+
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+
+    unique_watched = []
+    friend_titles_watched = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            # if movie["title"] in unique_watched:
+            #     unique_watched.pop(movie["title"])
+            friend_titles_watched.append(movie["title"])
+
+    for movie in user_data["watched"]:
+        if movie["title"] not in friend_titles_watched:
+            unique_watched.append(movie)
+
+    return unique_watched
+
+def get_friends_unique_watched(user_data):
+    friends_watched = []
+    friends_unique_watched = []
+    user_titles = set()
+
+    # iterate through user list
+       # append title to user_watched
+    for movie in user_data["watched"]:
+        user_titles.add(movie["title"])

```

Comment:
You can also convert the list returned from `user_data["watched"]` into a set by doing:
`user_titles = set(user_data["watched"])`

https://www.programiz.com/python-programming/set